modules/rostests/apitests/user32/ScrollBarRedraw.c

```diff
@@ -0,0 +1,712 @@
+    HaveHRedraw = TRUE;
+    HaveVRedraw = FALSE;
+    RunTestWindow(TEST_CLASS_NAME   "HRedraw",
+                  TEST_WINDOW_TITLE " (CS_HREDRAW)",
+                  CS_HREDRAW);
+
+    trace("Running test with CS_VREDRAW\n");
+    HaveHRedraw = FALSE;
+    HaveVRedraw = TRUE;
+    RunTestWindow(TEST_CLASS_NAME   "VRedraw",
+                  TEST_WINDOW_TITLE " (CS_VREDRAW)",
+                  CS_VREDRAW);
+
+    trace("Running test with both CS_HREDRAW and CS_VREDRAW\n");
+    HaveHRedraw = TRUE;
+    HaveVRedraw = TRUE;
+    RunTestWindow(TEST_CLASS_NAME   "HRedrawVRedraw",
+                  TEST_WINDOW_TITLE " (CS_HREDRAW | CS_VREDRAW)",
+                  CS_HREDRAW | CS_VREDRAW);
+
+    trace("Test complete\n");
+    ColorsCleanup();
+}
+
+static void HideVertScrollBar(HWND Window)
+{
+    SCROLLINFO ScrollInfo;
+
+    ScrollInfo.cbSize = sizeof(ScrollInfo);
+    ScrollInfo.fMask = SIF_RANGE | SIF_PAGE | SIF_POS;
+    ScrollInfo.nPage = ClientHeight;
+
+    ScrollInfo.nMin = 0;
+    ScrollInfo.nMax = ClientHeight - 1;
+    ScrollInfo.nPos = 0;
+
+    SetScrollInfo(Window, SB_VERT, &ScrollInfo, TRUE);
+}
+
+static void ShowVertScrollBar(HWND Window)
+{
+    SCROLLINFO ScrollInfo;
+
+    ScrollInfo.cbSize = sizeof(ScrollInfo);
+    ScrollInfo.fMask = SIF_RANGE | SIF_PAGE | SIF_POS;
+    ScrollInfo.nPage = ClientHeight;
+
+    ScrollInfo.nMin = 0;
+    ScrollInfo.nMax = (3 * ClientHeight) - 1;
+    ScrollInfo.nPos = 0;
+
+    SetScrollInfo(Window, SB_VERT, &ScrollInfo, TRUE);
+}
+
+static void HideHorzScrollBar(HWND Window)
+{
+    SCROLLINFO ScrollInfo;
+
+    ScrollInfo.cbSize = sizeof(ScrollInfo);
+    ScrollInfo.fMask = SIF_RANGE | SIF_PAGE | SIF_POS;
+    ScrollInfo.nPage = ClientWidth;
+
+    ScrollInfo.nMin = 0;
+    ScrollInfo.nMax = ClientWidth - 1;
+    ScrollInfo.nPos = 0;
+
+    SetScrollInfo(Window, SB_HORZ, &ScrollInfo, TRUE);
+}
+
+static void ShowHorzScrollBar(HWND Window)
+{
+    SCROLLINFO ScrollInfo;
+
+    ScrollInfo.cbSize = sizeof(ScrollInfo);
+    ScrollInfo.fMask = SIF_RANGE | SIF_PAGE | SIF_POS;
+    ScrollInfo.nPage = ClientWidth;
+
+    ScrollInfo.nMin = 0;
+    ScrollInfo.nMax = (3 * ClientWidth) - 1;
+    ScrollInfo.nPos = 0;
+
+    SetScrollInfo(Window, SB_HORZ, &ScrollInfo, TRUE);
+}
+
+static int FsmStep(HWND Window)
+{
+    static COLORREF PrevColor = CLR_INVALID;
+    COLORREF Color = CLR_INVALID;
+    HDC hdc = NULL;
+
+    if (FsmState != FSM_STATE_END)
+    {
+        hdc = GetDC(Window);
+
+        ok(hdc != NULL, "Failed to get device context\n");
+        if (hdc == NULL)
+        {
+            DestroyWindow(Window);
+            FsmState = FSM_STATE_END;
+        }
+        else
+        {
+            Color = GetPixel(hdc, ClientWidth / 4, ClientHeight / 4);
+            ok(Color != CLR_INVALID, "Failed to get window color\n");
+        }
+    }
+
+    trace("FsmState: %d, Color: 0x%.8lX\n", FsmState, Color);
+
+    switch (FsmState)
+    {
+        case FSM_STATE_START:
+
```

Comment:
Optional: I would encourage you to strip the empty lines after all those case labels. I will approve also without that optional sugeestion.